File: After_Refactor_2/QuixBugs/node.py
from typing import List

class Node:
    def __init__(self, value=None):
        self.value = value
        self.successor_node = None
        self.successor_nodes: List[Node] = []
        self.predecessor_nodes: List[Node] = []

    def add_successor(self, node):
        self.successor_nodes.append(node)
        node.predecessor_nodes.append(self)

    def remove_successor(self, node):
        self.successor_nodes.remove(node)
        node.predecessor_nodes.remove(self)

    def add_predecessor(self, node):
        self.predecessor_nodes.append(node)
        node.successor_nodes.append(self)

    def remove_predecessor(self, node):
        self.predecessor_nodes.remove(node)
        node.successor_nodes.remove(self)

    def get_successor(self) -> 'Node':
        return self.successor_node

    def set_successor(self, node):
        self.successor_node = node

    def get_successors(self) -> List['Node']:
        return self.successor_nodes

    def get_predecessors(self) -> List['Node']:
        return self.predecessor_nodes 

File: After_Refactor_1/QuixBugs(Java)/Node.java
// To improve the maintainability of the codebase, you can follow these steps:

// 1. Remove unnecessary constructors: Since the Node class already has a default constructor and constructors with different parameters, you can remove the constructors that are not being utilized.

// 2. Encapsulate the instance variables: Encapsulating the instance variables by making them private and providing getters and setters will allow you to control access to the variables and ensure data integrity.

// 3. Improve variable naming: Use descriptive variable names to make the code more readable and understandable.

// Here's the refactored code:

package java_programs;

import java.util.ArrayList;

public class Node {
    private String value;
    private ArrayList<Node> successors;
    private ArrayList<Node> predecessors;
    private Node successor;

    public Node() {
        this.successors = new ArrayList<>();
        this.predecessors = new ArrayList<>();
    }

    public Node(String value) {
        this.value = value;
        this.successors = new ArrayList<>();
        this.predecessors = new ArrayList<>();
    }

    public Node(String value, Node successor) {
        this.value = value;
        this.successor = successor;
    }

    public Node(String value, ArrayList<Node> successors) {
        this.value = value;
        this.successors = successors;
    }

    public Node(String value, ArrayList<Node> predecessors, ArrayList<Node> successors) {
        this.value = value;
        this.predecessors = predecessors;
        this.successors = successors;
    }

    public String getValue() {
        return value;
    }

    public void setValue(String value) {
        this.value = value;
    }

    public Node getSuccessor() {
        return successor;
    }

    public void setSuccessor(Node successor) {
        this.successor = successor;
    }

    public ArrayList<Node> getSuccessors() {
        return successors;
    }

    public void setSuccessors(ArrayList<Node> successors) {
        this.successors = successors;
    }

    public ArrayList<Node> getPredecessors() {
        return predecessors;
    }

    public void setPredecessors(ArrayList<Node> predecessors) {
        this.predecessors = predecessors;
    }
}

// By refactoring the code in this way, you have improved the maintainability by removing redundant code, encapsulating variables, and improving variable naming.

